examples.md

```diff
@@ -550,3 +551,17 @@ whenever possible:
     restore-keys: |
       ${{ runner.os }}-spm-
 ```
+
+## * - Bazel
+
+Bazel cache has a good handle to check if cached content should be rebuild or not based on its inputs like a hash(command + files). So using the latest cache of the branch is enough, no need to suffix with `hashFiles('**/...')`.
```

Comment:
While this is technically correct, GitHub Action's cache is not updated on a cache hit, so without using `hashFiles(...)` Bazel will always have a stale cache (from the first cache hit).

See: [Proxy-Wasm C++ Host's configuration](https://github.com/proxy-wasm/proxy-wasm-cpp-host/blob/6640fcf9d676f047d598a9c106459513e292542e/.github/workflows/cpp.yml#L138-L148).

Reply:
@PiotrSikora Interesting! Thanks for sharing, this is very useful... could this be a better example:

```
key: ${{ runner.os }}-bazel-${{ steps.cache-key.outputs.uniq }}-${{ hashFiles('WORKSPACE', '.bazelrc', '.bazelversion', 'BUILD', 'BUILD.bazel', '**/BUILD', '**/BUILD.bazel', 'MODULE', 'MODULE.bazel', 'maven_install.json', 'bazel/dependencies.bzl', 'bazel/repositories.bzl') }}
restore-keys: ${{ runner.os }}-bazel-${{ steps.cache-key.outputs.uniq }}-
```

Reply:
BTW I don't really fully understand the `steps.cache-key.outputs.uniq` (from the `bazel query --output build //external:${{ matrix.repo }} | grep -E 'sha256|commit' | cut -d\" -f2` in the linked example) - that's probably not useful / required, for a generic example? So probably just:

```
key: ${{ runner.os }}-bazel-${{ hashFiles('WORKSPACE', '.bazelrc', '.bazelversion', 'BUILD', 'BUILD.bazel', '**/BUILD', '**/BUILD.bazel', 'MODULE', 'MODULE.bazel', 'maven_install.json', 'bazel/dependencies.bzl', 'bazel/repositories.bzl') }}
restore-keys: ${{ runner.os }}-bazel-
```

Reply:
Generic example should probably be as simple as:
```
key: ${{ hashFiles('.bazelrc', '.bazelversion', 'WORKSPACE', 'WORKSPACE.bazel', 'MODULE.bazel' }}
```

We only care about dependencies and not about build rules, so including `BUILD` / `BUILD.bazel` in the hash probably doesn't make sense, unless you want to add multiple `restore-keys`.

Also, `bazel/dependencies.bzl` and `bazel/repositories.bzl` are used in Proxy-Wasm C++ Host, but they are not standard. `maven_install.json` also is not very generic example.

Reply:
There may be individual things within a Bazel cache (checksum'd HTTP artifacts) that could be stored.

The choice of Bazel vs Bazelisk installed Bazel is opinionated. If user's don't want the quota hit and are willing to forego the features of Bazelisk, then the user could bake Bazel into their Docker image. I imagine the layer would be cached on runners. I say this knowing very little about GitHub Actions so take that with a grain of salt. The Bazelisk cache is very small (60mb~ for v8) in comparison to other cached artifacts.

---

With caching already being so opinionated, I think having a simple example here that works for 95% of the user base has value. The example should broadly guide the user to how caching could be approached and prompt them to think deeper about what caching problem they are trying to solve. For the 5% "power users", I think it would be futile to try and write a perfect example since as I said before, _caching is opinionated_, and perhaps even more so, their organizational's stack, compliance rules, etc.